### DISCUSSION TITLE
How to know if a node is safe to delete?

### DISCUSSION THREAD

**Discussion by 1mike12** [mobxjs/mobx-state-tree#1853] [Q&A]
I don't see much discussion anywhere on how to best manage removal of objects in a safe manner when there are lots of relations. 

For instance, lets say we had a twitter clone. You would have an infinite scrolling list of tweets on the homescreen, then a map of users that a tweet will reference. We would want to delete tweets as we approach some threshold. We would also want to delete a user from the usermap if there's no tweets depending on it. 

How do we know if a certain user node is not referenced anymore? 

the only thing i found is this (with no answers) : https://spectrum.chat/mobx-state-tree/general/whats-the-best-way-to-track-references~29c39f2d-645d-43d7-9713-8aae7e6caad8 

Are we supposed to do reference counting ourselves? That seems really error prone

**Comment by coolsoftwaretyler** [comment]
Hey @1mike12 - I know this is a bit of an old thread, but does the section about reference invalidation in the docs help at all? 

https://mobx-state-tree.js.org//concepts/references#reference-validation-isvalidreference-tryreference-oninvalidated-hook-and-typessafereference

Could you use `onInvalidated` for something like this?

> The options parameter for references also accepts an optional onInvalidated hook, which will be called when the reference target node that the reference is pointing to is about to be detached/destroyed. It has the following signature:
